Add stat metrics semconv (#141)
coralogix/opentelemetry-ebpf-instrumentation: devdocs/stat-metrics-semconv.md

> # OTel Semantic Conventions Proposal — Stat Metrics (TCP/UDP)

Reviewer: I know this is an internal OBI proposal so this is fine
but when we propose anything to the semantic conventions we should not mention anything like `stats` or `statsolly` since these are internal OBI concepts

## Context

> The `statsolly` subsystem is where we capture **stat metrics**: numeric observations read directly from kernel state via eBPF hooks (kprobes, tracepoints, fentry, sockops). It is one of 3 pipelines the agent runs: statsolly, appoly, netolly.

Reviewer 2: ```suggestion
The `statso11y` subsystem is where we capture **stat metrics**: numeric observations read directly from the kernel via eBPF. It is one of 3 pipelines the agent runs: statso11y, appo11y, neto11y.
```

> Today, statsolly emits exactly 2 user-facing stat metrics:

Reviewer 2: I believe that for file names we use statsolly (with 'l') for simplicity, the correct name should be statso11y (stats o-bservabilit-y)

| Current name | Instrument | Unit | Source |
|---|---|---|---|
| `obi.stat.tcp.rtt` | Histogram | `s` | `kprobe/tcp_close` reading `tcp_sock.srtt_us` |

> | `obi.stat.tcp.failed.connections` | Counter | - | `sock/inet_sock_set_state` |

Reviewer 2: ```suggestion
| `obi.stat.tcp.failed.connections` | Counter | {connection} | `sock/inet_sock_set_state` |
```
should this be `{connection}` as per naming principles section?

Author: yep

There is no entry in OTel semantic conventions for these metrics, nor for any other TCP- or UDP-specific stat metric. The existing `system.network.*` namespace covers transport-agnostic NIC-level counters (`system.network.io`, `system.network.connection.count`, `system.network.packet.count`, `system.network.packet.dropped`, `system.network.errors`) but does not descend into protocol-specific detail.

We plan to expand statsolly's TCP and UDP coverage substantially. Before shipping those with new `obi.*` names we want to:

- retire the vendor prefix where possible
- align with existing OTel semantic conventions
- propose new semconv entries for what is not yet covered.

## Proposal

Add new TCP/UDP stat metrics to semconv by placing them under `system.network.*`, as sub-namespaces `system.network.tcp.*` and `system.network.udp.*`, alongside the existing transport-agnostic network metrics.

## Stat metrics vs OTel system metrics

> Stat metrics (as this project uses the term) and OTel's `system.*` metrics describe the same class of signal such as kernel/OS-observed telemetry, typically counters of events or histograms of observed quantities. The only difference is the collection mechanism: semconv's `system.*` was historically defined against procfs/sysfs polling (node_exporter shape); statsolly observes the same phenomena via eBPF.

Reviewer 2: not sure I 100% understood this section, maybe it can be made more clear or deleted if not necessary?

Author: I was pointing out the differences/similarities between what in obi we call stat metrics and what in otel is called "system.*" to make it clear why we want to put the metrics calculated by statsolly into "system.network.*".

Semantics, not the mechanism, define the namespace.

This proposal therefore positions statsolly's TCP/UDP additions as extensions to the existing `system.network.*` namespace, not as a parallel vendor space.

## Scope

> - TCP stat metrics via kprobes / tracepoints / more.

Reviewer 2: I think implementation details should be left out of this proposal, maybe we should put something like `TCP stat metrics: connection lifecycle, RTT, retransmits, resets, queue pressure, flow control` or similar

- UDP stat metrics (errors, queue drops, datagram sizing). Datagram *counts* reuse `system.network.packet.count{network.transport=udp}`.

> - A small set of new attributes strictly needed to describe the above.

Reviewer 2: ```suggestion
- New attributes required to describe the above.
```

> - A before -> after mapping for the already implemented stat metrics.

Reviewer 2: I don't think this is part of the scope

Author: do you mean only this part? "A before -> after mapping for the already implemented stat metrics"

Reviewer 2: yes

## Naming principles applied

From the semconv general metrics spec and existing precedent:

1. **Nest under `system.network.*`.** Existing stable semconv places network telemetry there; TCP/UDP sub-namespacing inside (`system.network.tcp.*`, `system.network.udp.*`) mirrors the `jvm.gc.*` / `jvm.memory.*` pattern of sub-namespacing within a stable root.
2. **No `.total` suffix** in OTel names (Prometheus exporter adds it).
3. **Units via UCUM**, not in the name: `By`, `s`, and grammatically-singular `{segment}`, `{connection}`, `{event}`, `{error}`, `{datagram}` for counts.

> 4. **Attributes over separate metrics** when aggregation across the attribute values is still a meaningful number. Separate metrics when events are semantically distinct (different hooks alone is not sufficient grounds to split).

Reviewer 2: I'd remove the last part (technical)

Author: which one? All list?

Reviewer 2: `(different hooks alone is not sufficient grounds to split).`

5. **Reuse stable attributes** (`error.type`, `network.io.direction`, `network.transport`, `network.connection.state`) before inventing new ones.

## Namespace layout

```
system.network.*                              — existing stable root
  system.network.io                           — existing
  system.network.packet.count                 — existing
  system.network.packet.dropped               — existing
  system.network.errors                       — existing
  system.network.connection.count             — existing

>   system.network.tcp.*                        — TCP-specific stat metrics (NEW)

Reviewer 2: IMHO we shouldn't use the internal "stat metrics" lingo for the public semconv proposal -- it's an internal term 

```suggestion
  system.network.tcp.*                        — TCP metrics (NEW)
```

Author: Done but this is not the proposal for the public semconv proposal it's for obi.

>     system.network.tcp.connection.successes   

Reviewer: maybe we can just track all connections and failures independently
and have
`system.network.tcp.connection.total`
`system.network.tcp.connection.errors`

and we can infer
successful connections = `system.network.tcp.connection.total - system.network.tcp.connection.errors`

Author: If we want to track all connections I think it should be harder to implement and we need to hook in a krpobe that is called for every SYN while now we are using a `sock/inet_sock_set_state` which is called during state transition.

Moreover, subtraction is noisy at query time. `total` and `errors` resolve at different moments and over any short window, `total - errors` includes unresolved in-flight attempts, so the derived "successes" wobbles. Explicit successes/failures counted at the terminal event doesn't have this problem.

    system.network.tcp.connection.failures   (now called obi.stat.tcp.failed.connections)
    system.network.tcp.connection.duration
    system.network.tcp.handshake.duration

>     system.network.tcp.rtt                   (now called obi.stat.tcp.rtt)

Reviewer: should this be `system.network.tcp.roundtrip.duration`?
since this is a hisgoram, even though rtt is a common concept

Author: I thing `system.network.tcp.rtt` or `system.network.tcp.rtt.duration` are the best choices because `rtt` is like a "standard"

Reviewer: `system.network.tcp.rtt.duration` sounds good

>     system.network.tcp.retransmits

Reviewer: maybe `system.network.tcp.retransmit.count`

Author: same thing as before.

>     system.network.tcp.resets

Reviewer: maybe `system.network.tcp.reset.count`

Author: Mmmh it should depend on the semantic of the metric. I saw both ways [`system.network.packet.count`](https://opentelemetry.io/docs/specs/semconv/system/system-metrics/#metric-systemnetworkpacketcount) and [`system.network.errors`](https://opentelemetry.io/docs/specs/semconv/system/system-metrics/#metric-systemnetworkerrors). I don't have a clear opinion on this.

Reviewer: im still not super sure, according to [otel](https://opentelemetry.io/docs/specs/semconv/general/naming/#pluralization)

> Metric namespaces SHOULD NOT be pluralized.
> 
> Metric names SHOULD NOT be pluralized, unless the value being recorded represents discrete instances of a [countable quantity](https://wikipedia.org/wiki/Count_noun). Generally, the name SHOULD be pluralized only if the unit of the metric in question is a non-unit (like {fault} or {operation}).

so i think all counters should be plural `system.network.tcp.resets` and all histograms should be single

Author: so let's do
- `system.network.tcp.resets`
- `system.network.tcp.retransmits`
- `system.network.udp.queue.drops`
WDYT?

>   system.network.udp.*                        — UDP-specific (NEW)

Reviewer 2: ```suggestion
  system.network.udp.*                        — UDP metrics (NEW)
```

    system.network.udp.errors

>     system.network.udp.queue.drops

Reviewer: `system.network.udp.queue.drop.count`

Author: check here https://github.com/coralogix/opentelemetry-ebpf-instrumentation/pull/141#discussion_r3117281219

    system.network.udp.datagram.size

  system.network.icmp.*                       — RESERVED, future proposal

>   system.network.quic.*                       — RESERVED, future proposal

Reviewer 2: I'd remove these for now

network.*                                     — attribute root (unchanged)
  network.transport                           — existing
  network.io.direction                        — existing
  network.connection.state                    — existing

>   network.tcp.handshake.role                  — NEW (client|server) or (local|remote)

Reviewer 2: should these be part of this proposal? If yes, I think we should set on names, eg. client/server (without or)

Author: I added it because I wanted to include other metrics in the proposal that haven't yet been implemented, to have a broader range. So yes, we'd have to decide between the two; I added them precisely to reach a solution during the review phase (the internal one or during the obi one)

  network.tcp.retransmit.type                 — NEW (fast|tail_loss_probe|timeout|spurious)
  network.tcp.reset.cause                     — NEW, optional (application|timeout|unreachable|refused)
```

Note that metric names sit under `system.network.*` while attribute names sit under `network.*` — matching the existing semconv layout (e.g., metric `system.network.connection.count` carries attribute `network.connection.state`).

## Proposed new metrics — TCP (implemented)

### `system.network.tcp.rtt`

- **Instrument**: Histogram
- **Unit**: `s`
- **Description**: Smoothed round-trip time (`srtt_us`) sampled at `tcp_close`. This supersedes `obi.stat.tcp.rtt`.
- **Attributes**: none required beyond resource attributes; may carry peer attributes where useful.

### `system.network.tcp.connection.failures`

- **Instrument**: Counter
- **Unit**: `{connection}`
- **Description**: TCP connection establishment attempts that did not reach `ESTABLISHED`.
- **Attributes**:
  - `error.type` — enum `unknown | refused | reset | timed-out | host-unreachable | net-unreachable | other` (reusing the stable `error.type` attribute used elsewhere in semconv). Now called `reason`.

>   - (not added till now) `network.tcp.handshake.role` (`client|server`)

Reviewer 2: ```suggestion
  - (NEW) `network.tcp.handshake.role` (`client|server`)
```

- **Attributes**: `network.tcp.flow_control.event` (`zero_window|window_full`), `network.io.direction`.

## Proposed new metrics — UDP

UDP is stateless: no connections, no retransmits, no handshake. Signals of interest are error categories and payload sizing.

### `system.network.udp.errors`

- **Instrument**: Counter
- **Unit**: `{error}`
- **Description**: UDP-level error events.
- **Attributes**: `error.type` — enum `checksum | port_unreachable | send_buffer | receive_buffer` (reusing stable `error.type`; `port_unreachable` chosen to match ICMP terminology users already know, rather than the Linux-specific `NoPort`). Note: `network.io.direction` where applicable.

### `system.network.udp.queue.drops`

- **Instrument**: Counter
- **Unit**: `{datagram}`
- **Description**: Datagrams dropped due to socket queue exhaustion (distinct from interface-level drops covered by `system.network.packet.dropped`).
- **Attributes**: `network.io.direction`.

### `system.network.udp.datagram.size`

- **Instrument**: Histogram
- **Unit**: `By`

> - **Description**: Datagram payload size distribution.

Reviewer 2: is this datagram size or datagram payload size? the metric name is a bit unclear

Author: Ops you are right. We can rename it to `system.network.udp.message.size`, in this way the description can be `Size distribution of UDP message payload.`.

In otel there is something similar https://opentelemetry.io/docs/specs/semconv/registry/attributes/messaging/#messaging-message-body-size which is `The size of the message body in bytes`.
WDYT?

Reviewer 2: as long as it's consistent with the metric name/description, it's fine

- **Attributes**: `network.io.direction`.

## Proposed new attributes

| Attribute | Values | Applies to |
|---|---|---|
| `network.tcp.handshake.role` | `client`, `server` | `system.network.tcp.connection.successes`, `system.network.tcp.connection.failures`, `system.network.tcp.connection.duration`, `system.network.tcp.handshake.duration` |
| `network.tcp.retransmit.type` | `fast`, `tail_loss_probe`, `timeout`, `spurious` | `system.network.tcp.retransmits` |
| `network.tcp.reset.cause` | `application`, `timeout`, `unreachable`, `refused` (optional) | `system.network.tcp.resets` |
| `network.tcp.flow_control.event` | `zero_window`, `window_full` | `system.network.tcp.flow_control.events` |

Attributes explicitly **reused** from the stable set: `error.type`, `network.io.direction`, `network.transport`, `network.connection.state`, `network.interface.name`.

> Related gap flagged for a separate proposal: `network.transport` does not currently include `sctp`.

Reviewer 2: not sure this is necessary in this proposal

> ## Before -> after mapping for this repo

Reviewer 2: do we need a migration phase? eg. by having the deprecated attribute plus the new one?

Author: This is a good question. Currently, the values ​​of `metrics.features` can be `stats_tcp_failed_connections` or `stats_tcp_rtt`, and if the proposal is accepted (at least internally), we could
1. change them to `system_network_tcp_connection_failures/system_network_tcp_rtt` or
2. leave the current values ​​since they are internal, thus avoiding a breaking change.

Reviewer 2: I think metrics.features can remain the same, I was talking about metrics

Author: Got it!

Reviewer 2: So what's the plan? :D should we document it?

Author: At least for the internal proposal I would leave this section "Before -> after mapping for this repo",for the otel-spec version we can remove it, it is not necessary.
wdyt?

Reviewer 2: for the otel proposal, this is not necessary -- for the OBI one I'd say it's important to have a "plan" for renaming/migrating the existing metrics

Author: Right! I can add a section "OBI rollout (repo-internal, not part of the semconv proposal)" and then 2 subsections "Before -> after mapping for this repo" and "Migration strategy (to be decided)" with a list of ideas.
